hcache/compr_lz4.c

```diff
@@ -0,0 +1,132 @@
+/**
+ * @file
+ * Header cache compression
+ *
+ * @authors
+ * Copyright (C) 2019 Tino Reichardt <milky-neomutt@mcmilk.de>
+ *
+ * @copyright
+ * This program is free software: you can redistribute it and/or modify it under
+ * the terms of the GNU General Public License as published by the Free Software
+ * Foundation, either version 2 of the License, or (at your option) any later
+ * version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/**
+ * @page LZ4 compression
+ *
+ * Use LZ4 header cache compression for database backends.
+ */
+
+#include "config.h"
+#ifdef HAVE_LZ4_H
+#include <stddef.h>
+#include <lz4.h>
+#include "mutt/mutt.h"
+#include "compr.h"
+#include "hcache.h"
+
+struct CCtx
+{
+  void *buf;
+};
+
+/**
+ * compr_lz4_open - Implements ComprOps::open()
+ */
+static void *compr_lz4_open(void)
+{
+  struct CCtx *ctx = mutt_mem_malloc(sizeof(struct CCtx));
+
+  ctx->buf = mutt_mem_malloc(LZ4_compressBound(1024 * 32));
+
+  return ctx;
+}
+
+/**
+ * compr_lz4_compress - Implements ComprOps::compress()
+ */
+static void *compr_lz4_compress(void *cctx, const char *data, size_t dlen, size_t *clen)
+{
+  struct CCtx *ctx = cctx;
+
+  if (!cctx || dlen < 10)
+    return NULL;
+
+  int datalen = dlen;
+  int len = LZ4_compressBound(dlen);
+  mutt_mem_realloc(&ctx->buf, len + 4);
+  char *cbuf = ctx->buf;
+
+  /* int LZ4_compress_fast (const char* src, char* dst, int srcSize, int dstCapacity, int acceleration); */
+  *clen = LZ4_compress_fast(data, cbuf + 4, datalen, len, C_HeaderCacheCompressLevel);
+  if (*clen < 0)
```

Comment:
`LZ4_compress_fast()` returns `int` -1 on failure, but `*clen` is a `size_t` so this `if` will never be true.

Reply:
It's a size_t no ssize_t ... ups :+1: